go/kbfs/libkbfs/subscription_manager.go

```diff
@@ -315,18 +310,127 @@ func (sm *subscriptionManager) SubscribePath(ctx context.Context,
+	return sm.cancelAndDeleteFolderBranchPollerLocked(sid)
+}
+
+func (sm *subscriptionManager) pollOnFolderBranchForSubscribePathRequest(
+	ctx context.Context, req subscribePathRequest, parsedPath *parsedPath) {
+	ticker := time.NewTicker(folderBranchPollingInterval)
+	for {
+		select {
+		case <-ticker.C:
+			fb, err := parsedPath.getFolderBranch(ctx, sm.config)
+			if err != nil {
+				_ = sm.cancelAndDeleteFolderBranchPoller(req.sid)
+				return
+			}
+
+			if fb == (data.FolderBranch{}) {
+				continue
+			}
+
+			// We have a folderBranch now! Go ahead and complete the
+			// sbuscription, and send a notification too.
+
+			sm.lock.Lock()
+			defer sm.lock.Unlock()
+			// Check if we're done while holding the lock to protect
+			// against racing against unsubscribe.
+			select {
+			case <-ctx.Done():
+				// No need to call cancelAndDeleteFolderBranchPollerLocked here
+				// since we always cancel and delete at the same tiem and if
+				// it's canceled it must have been deleted too.
+				return
+			default:
+			}
+
+			err = sm.subscribePathWithFolderBranchLocked(req, parsedPath, fb)
+			if err != nil {
+				sm.log.CErrorf(ctx,
+					"subscribePathWithFolderBranchLocked sid=%s err=%v", req.sid, err)
+			}
+
+			sm.notifier.OnPathChange(
+				sm.clientID, []SubscriptionID{req.sid},
+				req.path, []keybase1.PathSubscriptionTopic{req.topic})
+
+			_ = sm.cancelAndDeleteFolderBranchPollerLocked(req.sid)
+			return
+		case <-ctx.Done():
+			_ = sm.cancelAndDeleteFolderBranchPoller(req.sid)
+			return
+		}
+	}
+}
+
+func (sm *subscriptionManager) subscribePathWithoutFolderBranchLocked(
+	req subscribePathRequest, parsedPath *parsedPath) {
+	ctx, cancel := context.WithCancel(context.Background())
+	sm.folderBranchPollerCancelers[req.sid] = cancel
+	go sm.pollOnFolderBranchForSubscribePathRequest(ctx, req, parsedPath)
+}
+
+// SubscribePath implements the SubscriptionManager interface.
+func (sm *subscriptionManager) SubscribePath(ctx context.Context,
+	sid SubscriptionID, path string, topic keybase1.PathSubscriptionTopic,
+	deduplicateInterval *time.Duration) error {
+	sm.lock.Lock()
+	defer sm.lock.Unlock()
+
+	parsedPath, err := parsePath(userPath(path))
+	if err != nil {
+		return err
+	}
+	fb, err := parsedPath.getFolderBranch(ctx, sm.config)
+	if err != nil {
+		return err
+	}
+	req := subscribePathRequest{
+		sid:                 sid,
+		path:                path,
+		topic:               topic,
+		deduplicateInterval: deduplicateInterval,
+	}
+	if fb != (data.FolderBranch{}) {
+		return sm.subscribePathWithFolderBranchLocked(req, parsedPath, fb)
+	}
+	sm.subscribePathWithoutFolderBranchLocked(req, parsedPath)
+	return nil
+}
+
 // SubscribeNonPath implements the SubscriptionManager interface.
 func (sm *subscriptionManager) SubscribeNonPath(
 	ctx context.Context, sid SubscriptionID, topic keybase1.SubscriptionTopic,
 	deduplicateInterval *time.Duration) error {
+	// Lock at the beginnning to protect against racing with unsubscribe. We
```

Comment:
This comment is unrelated to the rest of this PR right?  That's fine, just making sure I understand.

Reply:
This is for this PR, but is supposed to be in SubscribePath instead of here, sorry! I'll explain in the other comment about locking.